### DISCUSSION TITLE
big.LITTLEのようなヘテロジニアスマルチプロセッサシステムへの対応について

### DISCUSSION THREAD

**Discussion by zacky1972** [toppers/users-forum#100] [General]
北九大の山崎進です．

最近，電力効率を向上させる目的で，big.LITTLEのような技術が普及しようとしています．

一方で，TOPPERS第3世代カーネル(ITRON系)統合仕様書のRelease 3.5.0を読みますと，TOPPERS/FMP3カーネルなどのマルチコアプロセッサ向けのカーネルはホモジニアスマルチプロセッサシステムが適用対象であると明記されています．

TOPPERSプロジェクトにおいて，big.LITTLEのようなヘテロジニアスマルチプロセッサシステムに対応しようという動きはあるのでしょうか？
もしあるのであれば，議論の状況を知りたいですし，何か貢献したいです．

**Comment by hiroakitakada** [comment]
統合仕様書では，以下のように記載しています。big.LITTLEのように，プロセッサの実装は異なっても，命令セットレベルで互換があるプロセッサで構成されたマルチコアは適用対象に含まれます。
```
(f) マルチプロセッサ対応カーネルにおいては，各プロセッサが同一の機械語
命令を実行できること【NGKI0003】．
```

命令セットレベルで互換性のないプロセッサで構成されたマルチコアの場合，カーネルのバイナリコードをコア毎に用意する必要があり，実装が大きく異なってきます。かなり以前に，学生の研究で，ARMとDSPで構成されたマルチコア向けにカーネルを開発したことがありますが，かなり大変でした。

**Comment by zacky1972** [comment]
高田先生，ありがとうございます．

もう少し正確に理解したいのですが，
TOPPERSカーネルで使用されている命令セットを，全てのプロセッサで問題なく実行できればOKという意味ですよね．

つまり，ユーザーアプリケーションレベルで使用する命令セットの一部が，big側に備わっているがLITTLEには備わっていないというような状況で，TOPPERSカーネルでこれらの命令を全く使用しないというような状況であれば，対象にできるという理解で合っていますか？

**Reply by hiroakitakada** [reply]
はい。その理解で正しいです。

**Reply by zacky1972** [reply]
ありがとうございます！
